File: datasets/C++/Qwen/DataStructures/hard4.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <climits>
using namespace std;

int main() {
    cout << "=== Find Maximum Si ===" << endl;

    int n;
    cout << "Enter the value of N: ";
    cin >> n;

    if (n < 2) {
        cout << "Maximum possible value of Si: 0" << endl;
        return 0;
    }

    vector<long long> A(n);
    for (int i = 0; i < n; i++) {
        cout << "Enter element " << i+1 << ": ";
        cin >> A[i];
    }

    vector<int> left(n, -1);
    vector<int> right(n, n);
    stack<int> st;

    // Find next smaller to the right
    for (int i = 0; i < n; i++) {
        while (!st.empty() && A[st.top()] > A[i]) {
            right[st.top()] = i;
            st.pop();
        }
        st.push(i);
    }

    while (!st.empty()) st.pop();

    // Find previous smaller to the left
    for (int i = n-1; i >= 0; i--) {
        while (!st.empty() && A[st.top()] > A[i]) {
            left[st.top()] = i;
            st.pop();
        }
        st.push(i);
    }

    long long max_si = 0;

    for (int i = 0; i < n; i++) {
        long long candidate = -1;  // smallest value > A[i] in range

        // Left side
        for (int j = i-1; j > left[i]; j--) {
            if (A[j] > A[i]) {
                if (candidate == -1 || A[j] < candidate) {
                    candidate = A[j];
                }
            }
        }

        // Right side
        for (int j = i+1; j < right[i]; j++) {
            if (A[j] > A[i]) {
                if (candidate == -1 || A[j] < candidate) {
                    candidate = A[j];
                }
            }
        }

        if (candidate != -1) {
            long long si = A[i] ^ candidate;
            if (si > max_si) {
                max_si = si;
            }
        }
    }

    cout << "Maximum possible value of Si: " << max_si << endl;
    return 0;
}

File: datasets/Python/Qwen/DataStructures/hard4.py
def main():
    print("=== Find Maximum Si ===")
    
    # Step 1: Get N
    n = int(input("Enter the value of N: "))
    
    if n < 2:
        print("Maximum possible value of Si: 0")
        return

    # Step 2: Get array elements one by one
    A = []
    for i in range(n):
        num = int(input(f"Enter element {i+1}: "))
        A.append(num)
    
    # Arrays to store left and right smaller element indices
    left = [-1] * n
    right = [n] * n
    stack = []

    # Find next smaller element to the right
    for i in range(n):
        while stack and A[stack[-1]] > A[i]:
            idx = stack.pop()
            right[idx] = i
        stack.append(i)

    stack = []
    
    # Find previous smaller element to the left
    for i in range(n-1, -1, -1):
        while stack and A[stack[-1]] > A[i]:
            idx = stack.pop()
            left[idx] = i
        stack.append(i)

    max_si = 0

    # For each element as minimum, find smallest greater element in its range
    for i in range(n):
        candidate = None

        # Check left side: from i-1 down to left[i] + 1
        for j in range(i-1, left[i], -1):
            if A[j] > A[i]:
                if candidate is None or A[j] < candidate:
                    candidate = A[j]

        # Check right side: from i+1 to right[i] - 1
        for j in range(i+1, right[i]):
            if A[j] > A[i]:
                if candidate is None or A[j] < candidate:
                    candidate = A[j]

        # If a valid second smallest exists
        if candidate is not None:
            si = A[i] ^ candidate
            if si > max_si:
                max_si = si

    print(f"Maximum possible value of Si: {max_si}")

if __name__ == "__main__":
    main()